lib/rust/ensogl/component/slider/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+
+            drag_pos_start          <- mouse.position.sample(&component_click);
+            drag_pos_end            <- mouse.position.gate(&component_drag);
+            drag_pos_end            <- any2(&drag_pos_end, &drag_pos_start);
+            drag_delta              <- all2(&drag_pos_end, &drag_pos_start).map(|(end, start)| end - start);
+
+            mouse_y_local           <- mouse.position.map(
+                f!([scene, model] (pos) scene.screen_to_object_space(&model.background, *pos).y )
+            ).gate(&component_drag);
+
+
+
+            // Componenet size
+
+            output.width            <+ input.set_width;
+            output.height           <+ input.set_height;
+
+
+            // precision calculation
+
+            precision_adjust_margin <- all2(&input.set_height, &input.set_precision_step_margin).map(
+                |(height, margin)| height / 2.0 + margin
+            );
+            precision_adjusted      <- all4(
+                &input.set_precision,
+                &mouse_y_local,
+                &precision_adjust_margin,
+                &input.set_precision_step_size,
+            ).map(
+                |(default, offset, margin, step_size)| {
+                    let steps = ((offset.abs() - margin) / step_size).max(0.0).ceil() * offset.signum();
+                    *default * (PRECISION_STEP_BASE).pow(steps)
+                }
+            );
+
+
+            // value calculation
+
+            value_reset             <- input.set_value_default.sample(&component_ctrl_click);
+            value_start             <- output.value.sample(&component_click);
+            value_start             <- any2(&value_reset, &value_start);
+
+            value_update            <- bool(&component_release, &value_start); // update only after value_start is sampled
+            value                   <- all3(&value_start, &precision_adjusted, &drag_delta).map(
+                |(value, precision, delta)| value + delta.x * precision
+            ).gate(&value_update);
+            value                   <- any2(&input.set_value, &value);
+            value                   <- all2(&value, &precision_adjusted).map(
+                |(value, precision)| (value / precision).round() * precision
+            );
+            value                   <- all3(&value, &input.set_value_min, &input.set_value_max).map(
+                |(value, min, max)| value.clamp(*min, *max)
+            );
+            output.value            <+ value;
+
+            track_pos.target        <+ all3(&value, &input.set_value_min, &input.set_value_max).map(
+                |(value, min, max)| (value - min) / (max - min)
+            );
+
+
+            // model update
+
+            eval track_pos.value (
+                (v) model.track.value.set(*v);
+            );
+            component_size <- all2(&input.set_width, &input.set_height).map(
+                |(width, height)| Vector2::new(*width, *height)
+            );
```

Comment:
This expression can be written in single line as well. We allow (see style guide) short var names if they are "obvious" - if they are explained in the same line. This can be written as 
```
component_size <- all2(&input.set_width, &input.set_height).map(|(w, h)| Vector2(*w, *h));
```

(See that there is no `::new` after `Vector`.